File: src/config/schema.ts
/**
 * Configuration schema — defines the shape of .sentinel/config.toml
 *
 * Uses Zod for runtime validation so we can give clear error messages
 * when config is malformed.
 */

import { z } from "zod";

export const ProviderNameSchema = z.enum(["claude", "openai", "gemini", "local"]);
export type ProviderName = z.infer<typeof ProviderNameSchema>;

export const RoleNameSchema = z.enum([
  "monitor",
  "researcher",
  "planner",
  "coder",
  "reviewer",
]);
export type RoleName = z.infer<typeof RoleNameSchema>;

export const ROLE_DESCRIPTIONS: Record<RoleName, string> = {
  monitor:
    "Scans your codebase continuously. Detects drift, tracks changes, assesses state. Runs often — should be cheap.",
  researcher:
    "Deep research, web search, reads docs and papers. Evaluates alternatives and best practices. Needs web access and long context.",
  planner:
    "Makes strategic decisions. Prioritizes work, decomposes tasks, sets architecture direction. Needs the best judgment available.",
  coder:
    "Writes code, runs tests, executes the plan. Needs agentic tool use (file editing, terminal, self-correction).",
  reviewer:
    "Verifies completed work. Code review, acceptance criteria checks. Should ideally be a different provider than coder for independence.",
};

export const ROLE_DEFAULTS: Record<RoleName, { provider: ProviderName; model: string }> = {
  monitor: { provider: "local", model: "qwen2.5-coder:14b" },
  researcher: { provider: "gemini", model: "gemini-2.5-pro" },
  planner: { provider: "claude", model: "claude-opus-4-6" },
  coder: { provider: "claude", model: "claude-sonnet-4-6" },
  reviewer: { provider: "gemini", model: "gemini-2.5-pro" },
};

export const RoleConfigSchema = z.object({
  provider: ProviderNameSchema,
  model: z.string(),
  endpoint: z.string().optional(),
  webSearch: z.boolean().optional(),
  thinking: z.boolean().optional(),
  maxContext: z.number().optional(),
  mode: z.enum(["agent-sdk", "api", "local"]).optional(),
  maxBudgetPerTaskUsd: z.number().optional(),
  schedule: z.enum(["on-change", "hourly", "daily", "manual"]).optional(),
});
export type RoleConfig = z.infer<typeof RoleConfigSchema>;

export const BudgetConfigSchema = z.object({
  dailyLimitUsd: z.number().default(15),
  warnAtUsd: z.number().default(10),
  trackLocalTokens: z.boolean().default(true),
});
export type BudgetConfig = z.infer<typeof BudgetConfigSchema>;

export const LocalConfigSchema = z.object({
  ollamaEndpoint: z.string().default("http://localhost:11434"),
  modelsPulled: z.array(z.string()).default([]),
  autoStartOllama: z.boolean().default(true),
});
export type LocalConfig = z.infer<typeof LocalConfigSchema>;

export const GoalsSchema = z.object({
  description: z.string(),
  milestones: z.array(z.string()).default([]),
  constraints: z.array(z.string()).default([]),
  priorities: z.array(z.string()).default([]),
});
export type Goals = z.infer<typeof GoalsSchema>;

export const SentinelConfigSchema = z.object({
  project: z.object({
    name: z.string(),
    path: z.string(),
  }),
  goals: GoalsSchema.optional(),
  roles: z.object({
    monitor: RoleConfigSchema,
    researcher: RoleConfigSchema,
    planner: RoleConfigSchema,
    coder: RoleConfigSchema,
    reviewer: RoleConfigSchema,
  }),
  budget: BudgetConfigSchema.default({}),
  local: LocalConfigSchema.default({}),
  research: z
    .object({
      multiModelConsensus: z.boolean().default(false),
      maxSourcesPerQuery: z.number().default(10),
      cacheTtlMinutes: z.number().default(60),
    })
    .default({}),
});
export type SentinelConfig = z.infer<typeof SentinelConfigSchema>;

File: src/sentinel/config/schema.py
"""
Configuration schema — defines the shape of .sentinel/config.toml

Uses Pydantic for runtime validation so we can give clear error messages
when config is malformed.
"""

from __future__ import annotations

from enum import StrEnum
from typing import Literal

from pydantic import BaseModel, Field


class ProviderName(StrEnum):
    CLAUDE = "claude"
    OPENAI = "openai"
    GEMINI = "gemini"
    LOCAL = "local"


class RoleName(StrEnum):
    MONITOR = "monitor"
    RESEARCHER = "researcher"
    PLANNER = "planner"
    CODER = "coder"
    REVIEWER = "reviewer"


ROLE_DESCRIPTIONS: dict[RoleName, str] = {
    RoleName.MONITOR: (
        "Scans your codebase continuously. Detects drift, tracks changes, "
        "assesses state. Runs often — should be cheap."
    ),
    RoleName.RESEARCHER: (
        "Deep research, web search, reads docs and papers. Evaluates "
        "alternatives and best practices. Needs web access and long context."
    ),
    RoleName.PLANNER: (
        "Makes strategic decisions. Prioritizes work, decomposes tasks, "
        "sets architecture direction. Needs the best judgment available."
    ),
    RoleName.CODER: (
        "Writes code, runs tests, executes the plan. Needs agentic tool use "
        "(file editing, terminal, self-correction)."
    ),
    RoleName.REVIEWER: (
        "Verifies completed work. Code review, acceptance criteria checks. "
        "Should ideally be a different provider than coder for independence."
    ),
}


class RoleDefault(BaseModel):
    provider: ProviderName
    model: str


ROLE_DEFAULTS: dict[RoleName, RoleDefault] = {
    RoleName.MONITOR: RoleDefault(provider=ProviderName.LOCAL, model="qwen2.5-coder:14b"),
    RoleName.RESEARCHER: RoleDefault(provider=ProviderName.GEMINI, model="gemini-2.5-pro"),
    RoleName.PLANNER: RoleDefault(provider=ProviderName.CLAUDE, model="claude-opus-4-6"),
    RoleName.CODER: RoleDefault(provider=ProviderName.CLAUDE, model="claude-sonnet-4-6"),
    RoleName.REVIEWER: RoleDefault(provider=ProviderName.GEMINI, model="gemini-2.5-pro"),
}


class RoleConfig(BaseModel):
    provider: ProviderName
    model: str
    endpoint: str | None = None
    web_search: bool = False
    thinking: bool = False
    max_context: int | None = None
    mode: Literal["agent-sdk", "api", "local"] | None = None
    max_budget_per_task_usd: float | None = None
    schedule: Literal["on-change", "hourly", "daily", "manual"] = "manual"


class Goals(BaseModel):
    description: str
    milestones: list[str] = Field(default_factory=list)
    constraints: list[str] = Field(default_factory=list)
    priorities: list[str] = Field(default_factory=list)


class BudgetConfig(BaseModel):
    daily_limit_usd: float = 15.0
    warn_at_usd: float = 10.0
    track_local_tokens: bool = True


class LocalConfig(BaseModel):
    ollama_endpoint: str = "http://localhost:11434"
    models_pulled: list[str] = Field(default_factory=list)
    auto_start_ollama: bool = True


class ResearchConfig(BaseModel):
    multi_model_consensus: bool = False
    max_sources_per_query: int = 10
    cache_ttl_minutes: int = 60


class RolesConfig(BaseModel):
    monitor: RoleConfig
    researcher: RoleConfig
    planner: RoleConfig
    coder: RoleConfig
    reviewer: RoleConfig


class ProjectConfig(BaseModel):
    name: str
    path: str


class SentinelConfig(BaseModel):
    project: ProjectConfig
    goals: Goals | None = None
    roles: RolesConfig
    budget: BudgetConfig = Field(default_factory=BudgetConfig)
    local: LocalConfig = Field(default_factory=LocalConfig)
    research: ResearchConfig = Field(default_factory=ResearchConfig)
